pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1053 @@
+// Copyright 2021 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge Chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+#![allow(clippy::map_identity)]
+
+use cfg_traits::{
+	investments::{ForeignInvestment, Investment, InvestmentCollector},
+	SimpleCurrencyConversion, StatusNotificationHook, TokenSwaps,
+};
+use cfg_types::investments::{
+	CollectedAmount, ExecutedForeignCollectInvest, ExecutedForeignCollectRedeem,
+	ExecutedForeignDecrease, Swap,
+};
+use frame_support::{ensure, traits::Get, transactional};
+use sp_runtime::{
+	traits::{EnsureAdd, EnsureAddAssign, Zero},
+	DispatchError, DispatchResult,
+};
+
+use crate::{
+	errors::{InvestError, RedeemError},
+	types::{
+		InnerRedeemState, InvestState, InvestTransition, RedeemState, RedeemTransition,
+		TokenSwapReason,
+	},
+	CollectedRedemptionTrancheTokens, Config, Error, Event, ForeignInvestmentInfo,
+	ForeignInvestmentInfoOf, InvestmentState, Pallet, RedemptionState, SwapOf, TokenSwapOrderIds,
+};
+
+mod invest;
+mod redeem;
+
+// Hook execution for (partially) fulfilled token swaps which should be consumed
+// by `TokenSwaps`.
+impl<T: Config> StatusNotificationHook for Pallet<T> {
+	type Error = DispatchError;
+	type Id = T::TokenSwapOrderId;
+	type Status = SwapOf<T>;
+
+	fn notify_status_change(
+		id: T::TokenSwapOrderId,
+		status: SwapOf<T>,
+	) -> Result<(), DispatchError> {
+		let info = ForeignInvestmentInfo::<T>::get(id).ok_or(Error::<T>::InvestmentInfoNotFound)?;
+		let reason = info
+			.last_swap_reason
+			.ok_or(Error::<T>::TokenSwapReasonNotFound)?;
+
+		match reason {
+			TokenSwapReason::Investment => {
+				let pre_state = InvestmentState::<T>::get(&info.owner, info.id).unwrap_or_default();
```

Comment:
When would we hit the `default()`?

Reply:
I just refactored the `InvestmentState` and `RedemptionState` storages to be `ValueQuery`. So they default to `NoState` in case it has never existed or was previously cleared. In the code you reviewed, it also defaulted to that.